pkg/updates/memory.go

```diff
@@ -0,0 +1,188 @@
+)
+
+// EdgeSet maps a node id to a list of node ids that it can update to
+type EdgeSet map[string][]string
+
+// NodeSet maps a node id to an index in the OrderedNodes array
+type NodeSet map[string]int
+
+// MemorySource is an in-memory implementation of Source.
+// It's an oracle to answer update questions for an installed version.
+type MemorySource struct {
+	// OrderedNodes is an ordered list of all nodes. Lower index == newer version.
+	OrderedNodes []State
+	// Nodes is a helper to lookup a node by id
+	Nodes NodeSet
+	// Edges contains the edgeset for this source.
+	Edges EdgeSet
+}
+
+// Next returns the newest version that can be installed in one step.
+func (m *MemorySource) Next(from string) string {
+	if edges, ok := m.Edges[from]; ok && len(edges) > 0 {
+		return edges[len(edges)-1]
+	}
+	return ""
+}
+
+// NextDirect returns the newest version that can be directly installed without
+// running any migrations.
+func (m *MemorySource) NextDirect(from string) (found string) {
+	initial := m.OrderedNodes[m.Nodes[from]]
+	if to, ok := m.Edges[from]; ok && len(to) > 0 {
+		for _, n := range m.Edges[from] {
+			node := m.OrderedNodes[m.Nodes[n]]
+
+			// if the phase and migration match the current node, no migrations
+			// are required
+			if initial.Phase == node.Phase && initial.Migration == node.Migration {
+				found = n
+			} else {
+				break
+			}
+		}
+	}
+	return found
+}
+
+// Latest returns the newest version that can be installed. If different
+// from `Next`, that means multiple steps are required (i.e. a multi-phase
+// migration, or a required stopping point in a series of updates).
+func (m *MemorySource) Latest(id string) string {
+	if len(m.OrderedNodes) == 0 || id == m.OrderedNodes[0].ID {
+		return ""
+	}
+	return m.OrderedNodes[0].ID
+}
+
+func (m *MemorySource) State(id string) State {
+	index, ok := m.Nodes[id]
+	if !ok {
+		return State{}
+	}
+	return m.OrderedNodes[index]
+}
+
+// TODO: consider *State instead of State
+func (m *MemorySource) Source(to string) (Source, error) {
+	// copy the ordered node list from `to` onward
+	var index int
+	if len(to) > 0 {
+		index = m.Nodes[to]
+	}
+	orderedNodes := make([]State, len(m.OrderedNodes)-index)
+	copy(orderedNodes, m.OrderedNodes[index:len(m.OrderedNodes)])
+
+	nodeSet := make(map[string]int, len(orderedNodes))
+	for i, n := range orderedNodes {
+		nodeSet[n.ID] = i
+	}
+
+	edges := make(map[string][]string)
+	for from, to := range m.Edges {
+		// skip edges where from is not in the node set
+		if _, ok := nodeSet[from]; !ok {
+			continue
+		}
+		_, ok := edges[from]
+		if !ok {
+			edges[from] = make([]string, 0)
+		}
+		for _, n := range to {
+			// skip edges where to is not in the node set
+			if _, ok := nodeSet[n]; !ok {
+				continue
+			}
+			edges[from] = append(edges[from], n)
+		}
+	}
+
+	return newMemorySourceFromValidatedNodes(nodeSet, edges, orderedNodes)
+}
+
+func (m *MemorySource) Matches(tag, digest, migration, phase string) []string {
```

Comment:
`FindMatchesFor`?

Reply:
ah, didn't end up using this, I'll just remove it.